runtime/oti/ObjectMonitor.hpp

```diff
@@ -473,6 +481,113 @@ class VM_ObjectMonitor
+		 */
+		UDATA lockReservationHistory = clazz->lockReservationHistory;
+
+		/* If incrementing the Cancel Counter would cause an overflow, first divide Cancel Counter and Reserved Counter by 2. */
+		if ((lockReservationHistory & 0xFFFF) < 0xFFFF) {
+			clazz->lockReservationHistory = lockReservationHistory + 1;
+		} else {
+			clazz->lockReservationHistory = ((lockReservationHistory >> 1) & 0x7FFF7FFF) + 1;
+		}
+	}
+
+	/**
+	 * Determine initial lockword value based on lockReservationHistory stored in the J9Class.
+	 *
+	 * @param javaVM[in] the J9JavaVM
+	 * @param clazz[in] the object's J9Class
+	 *
+	 * @returns The initial lockword value for the object
+	 */
+	static VMINLINE j9objectmonitor_t
+	getInitialLockword(J9JavaVM *javaVM, J9Class *clazz)
+	{
+		j9objectmonitor_t initialLockword = 0;
+
+		if (javaVM->enableGlobalLockReservation) {
+			/* This path is taken when Global Lock Reservation is enabled. */
+			UDATA lockReservationHistory = clazz->lockReservationHistory;
+
+			/*
+			 * Extract reservedCount and cancelCount from lockReservationHistory.
+			 * cancelCount is the lowest 16 bits of lockReservationHistory.
+			 * reservedCount is the second lowest 16 bits of lockReservationHistory.
+			 *
+			 * lockReservationHistory Format:
+			 * | Unused alignment padding (0/32 bits) | Reserved Counter (Unsigned 16 bits) | Cancel Counter (Unsigned 16 bits) |
+			 */
+			U_32 reservedCount = (lockReservationHistory >> 16) & 0xFFFF;
+			U_32 cancelCount = lockReservationHistory & 0xFFFF;
+
+			U_32 reservedAbsoluteThreshold = javaVM->reservedAbsoluteThreshold;
+			U_32 minimumReservedRatio = javaVM->minimumReservedRatio;
+
+			U_32 cancelAbsoluteThreshold = javaVM->cancelAbsoluteThreshold;
+			U_32 minimumLearningRatio = javaVM->minimumLearningRatio;
+
+			/*
+			 * Check reservedCount and cancelCount against different thresholds to determine what to initialize the lockword to.
+			 * First check if the ratio of resevation to cancellations is high enough to set the lockword to the New-AutoReserve state.
+			 * If so, the initial lockword value is OBJECT_HEADER_LOCK_RESERVED.
+			 * Second check if the ratio is high enough to set the lockword to the New-PreLearning state.
+			 * If so, the initial lockword value is OBJECT_HEADER_LOCK_LEARNING.
+			 * If the ratio is too low, the lockword starts in the Flat-Unlocked state and 0 is returned for the lockword value.
+			 *
+			 * Start as New-AutoReserve -> Initial lockword: OBJECT_HEADER_LOCK_RESERVED
+			 * Start as New-PreLearning -> Initial lockword: OBJECT_HEADER_LOCK_LEARNING
+			 * Start as Flat-Unlocked   -> Initial lockword: 0
+			 */
+			if ((reservedCount >= reservedAbsoluteThreshold) && (reservedCount > (cancelCount * minimumReservedRatio))) {
+				initialLockword = OBJECT_HEADER_LOCK_RESERVED;
+			} else if ((cancelCount < cancelAbsoluteThreshold) || (reservedCount > (cancelCount * minimumLearningRatio))) {
+				initialLockword = OBJECT_HEADER_LOCK_LEARNING;
+			}
```

Comment:
I felt that always accessing the two sub-fields together increases the chance of value drifting (i.e.  multi-threads trash each other's updates).  By and large, each routine mostly increments (modifies) one sub-field, except under overflow conditions (very very rare ... we can ignore). Accessing one sub-field would avoid potential trashing mostly.

Reply:
One minor side-effect of accessing the whole UDATA:  the data field locations are different in memory image on Big-Endian and Little-Endian platforms.  Likely it doesn't matter. Consumability?

Reply:
I'm making a change to separate `reservedCounter` and `cancelCounter` into two `U_16` values in `J9Class`. This means that they are now updated independently of each other. So incrementing one counter will not touch the other. However, this will also mean when handling overflow at the point where the counters are shifted right by 1 (to divide them in half) it will be possible to read in between updating the two counters.

Reply:
Note there is no implicit alignment in J9Class, so you can't have U16/U16/UDATA. Either the U16s will have to take the place of an existing U32, or U32 padding will need to be added on 64-bit platforms.

Pretty much all of our structures assume that their fields are aligned to their type - this is especially important for fields that we use atomics on.

Reply:
I was planning to do something like this:
```
union {
	UDATA padding; // This is used to preserve alignment under 64 bit.
	struct {
		U_16 reservedCounter;
		U_16 cancelCounter;
	} counters;
} lockReservationHistory;
```
The UDATA should preserve the alignment while adding in the two 16bit counters. Will something like that work?

Reply:
I don't see the need to go that complicated.  for 64bit, go with two U_32;  for 32bit, go with two U_16.  Either can be used to count to top of 0xFFFF.  Any drawback of this?

Reply:
Personally, I prefer the explicit ifdef padding, so that's it's obvious (we don't want to start accumulating lots of these unions that waste padding space).

Reply:
Having the fields be different types will be a pain for both the C code and the JIT code (you'll be making assumptions about the field size under ifdef).

Reply:
I decided to go with putting in an explicit ifdef to handle padding. Changes are in the new commit I just added.